### PR TITLE
Update trainer for easier handling of accumulate, compile fixes, and proper reporting

### LINKED ISSUE #34402: Accelerate + Dynamo broken in 4.46.0 due to model loss functions refactor
### System Info

- `transformers` version: 4.46.0
- Platform: Linux-5.15.0-91-generic-x86_64-with-glibc2.35
- Python version: 3.9.16
- Huggingface_hub version: 0.24.0
- Safetensors version: 0.4.5
- Accelerate version: 1.0.1
- Accelerate config:    - compute_environment: LOCAL_MACHINE
        - distributed_type: NO
        - mixed_precision: bf16
        - use_cpu: False
        - debug: False
        - num_processes: 1
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: 0
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: True
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
        - dynamo_config: {'dynamo_backend': 'INDUCTOR'}
- PyTorch version (GPU?): 2.5.0+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No
- Using GPU in script?: Yes
- GPU type: NVIDIA RTX A6000

### Who can help?

@muellerzr @ArthurZucker

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

#34191 introduced custom loss functions to the model classes. This appears to have broken training with accelerate + torch dynamo.

To reproduce, use `run_clm.py` with the following accelerate config:
```yaml
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: 'NO'
downcast_bf16: 'no'
dynamo_config:
  dynamo_backend: INDUCTOR
enable_cpu_affinity: true
gpu_ids: '0'
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 1
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```
```bash
accelerate launch run_clm.py \
    --log_level info \
    --model_name_or_path=meta-llama/Llama-3.2-1B \
    --dataset_name=Salesforce/wikitext \
    --dataset_config_name=wikitext-2-raw-v1 \
    --block_size=1024 \
    --per_device_train_batch_size=4 \
    --do_train \
    --bf16 \
    --output_dir=Llama-3.2-1B-wikitext-2-raw-v1 \
    --overwrite_output_dir \
    --seed=42 \
    --logging_steps=10 \
    --lr_scheduler_type=cosine \
    --num_train_epochs=3 \
    --learning_rate=5e-05 \
    --warmup_ratio=0.03 \
    --dataloader_drop_last
```

This produces an error from dynamo relating to the new `model_cls.loss_function` attribute added in #34191:
https://github.com/huggingface/transformers/blob/239a256a0c3bf2edab2bbd614923c4eaf88b867d/src/transformers/models/llama/modeling_llama.py#L1209-L1211

Important part of the traceback:
```text
  File "/anaconda3/envs/dev/lib/python3.9/site-packages/torch/_dynamo/variables/functions.py", line 152, in __init__
    assert isinstance(
AssertionError: expected FunctionType found _lru_cache_wrapper <functools._lru_cache_wrapper object at 0x7f1091109a40>

from user code:
   File "/anaconda3/envs/dev/lib/python3.9/site-packages/accelerate/utils/operations.py", line 820, in forward
    return model_forward(*args, **kwargs)
  File "/anaconda3/envs/dev/lib/python3.9/site-packages/accelerate/utils/operations.py", line 808, in __call__
    return convert_to_fp32(self.model_forward(*args, **kwargs))
  File "/anaconda3/envs/dev/lib/python3.9/site-packages/torch/amp/autocast_mode.py", line 44, in decorate_autocast
    return func(*args, **kwargs)
  File "/anaconda3/envs/dev/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py", line 1214, in forward
    loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.vocab_size, **loss_kwargs)

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
```

Now, if you update the accelerate config to not use dynamo, it runs just fine:
```yaml
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: 'NO'
downcast_bf16: 'no'
enable_cpu_affinity: true
gpu_ids: '0'
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 1
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

### Expected behavior

Accelerate should not throw the error when using torch dynamo.

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

Alternative to https://github.com/huggingface/transformers/pull/34442

TL;DR we just need to remove `lru_cache` and everything will work fine. (and adds a test)

This PR also takes the full lessons from my article and adds it to the `Trainer` for a simpler solution to the grad accum calculation (we shouldn't rely on `accelerator` from now on bc it can't handle the nuances with the grad accum fix at the highest level API, so we use a lower level version instead)

Fixes https://github.com/huggingface/transformers/issues/34402

**Would recommend a patch after this**

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @Rocketknight1

### CODE REVIEW

File: src/transformers/testing_utils.py [hunk 1]
```diff
@@ -1292,6 +1293,16 @@ def require_jumanpp(test_case):
     return unittest.skipUnless(is_jumanpp_available(), "test requires jumanpp")(test_case)
 
 
+def require_python_311(test_case):
```
Comment (Rocketknight1) [inline src/transformers/testing_utils.py:1296]: Do we need the decorator? The test passes on Python < 3.11 anyway, so we can simplify the codebase and run it regardless!

**Review by Rocketknight1 — COMMENTED** [review]
Couple comments about the test!

**Review by Rocketknight1 — APPROVED** [review]
Tests look clean to me now, and I'm trusting you on the accelerate side of things! :sweat_smile:

cc @LysandreJik / @ArthurZucker for core maintainer review

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -2491,7 +2490,13 @@ def _inner_training_loop(
                     if step % args.gradient_accumulation_steps == 0:
                         self.control = self.callback_handler.on_step_begin(args, self.state, self.control)
 
-                    with self.accelerator.accumulate(model):
+                    # We explicitly want to avoid relying on `accelerator.accumulate` for generation training
+                    context = (
+                        functools.partial(self.accelerator.no_sync, model=model)
+                        if i == len(batch_samples) - 1
+                        else contextlib.nullcontext
+                    )
```
Comment (muellerzr) [inline src/transformers/trainer.py:2498]: For an explanation on what we have going on here @Rocketknight1 , during DDP we use `model.no_sync()` to only communicate across all GPUs during the next step outside it (so we speed up training when not needed when doing gradient accumulation). `accelerator.no_sync()` is the lower-level `accumulate()` API which makes that op backed-independent (so on a single GPU it just does `nullcontext`)

**Review by Milad335t — APPROVED** [review]

**Review by Milad335t — APPROVED** [review]

**Review by Milad335t — APPROVED** [review]

**Review by Milad335t — APPROVED** [review]
M

**Review by Milad335t — APPROVED** [review]

**Comment by ArthurZucker** [comment]
@Milad335t just warning you to stop spamming or we'll have to block you 😢

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -601,8 +600,8 @@ def __init__(
             if not _is_peft_model(unwrapped_model)
             else unwrapped_model.get_base_model().forward
         )
-
-        self.model_accepts_loss_kwargs = "loss_kwargs" in inspect.signature(model_forward).parameters
+        params = inspect.signature(model_forward).parameters
+        self.model_accepts_loss_kwargs = "loss_kwargs" in params or "kwargs" in params
```
Comment (ArthurZucker) [inline src/transformers/trainer.py]: works, checking for `ParameterKind.VAR_KEYWOR` as type is probably better

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -5100,9 +5097,10 @@ def get_batch_samples(self, epoch_iterator, num_batches):
         if len(batch_samples) > 0 and "labels" in batch_samples[0]:
             # For now we don't support object detection
             try:
-                num_items_in_batch = sum(
-                    [data_batch["labels"][..., 1:].ne(-100).sum().item() for data_batch in batch_samples]
-                )
+                num_items_in_batch = sum([(batch["labels"].ne(-100)).sum() for batch in batch_samples])
```
Comment (ArthurZucker) [inline src/transformers/trainer.py:5099]: weird to me that we have to use -100 here, instead of a general parameter but whit was already the case

**Review by ArthurZucker — COMMENTED** [review]
Thanks, let's hope this gets stabilized!

File: src/transformers/trainer.py (hunk 4 above)
Comment (muellerzr) [reply src/transformers/trainer.py:5104]: IIRC we use `-100` for padding by default in the Trainer. I can align it to `self.processor` if it exists else `-100` if that's better?:)

Comment (ArthurZucker) [reply src/transformers/trainer.py:5107]: okay sounds good then sorry

**Review by ArthurZucker — APPROVED** [review]
Thanks patching today!

File: src/transformers/trainer.py (hunk 4 above)
Comment (muellerzr) [reply src/transformers/trainer.py:5107]: No worries, it's weird for me too :)

Comment (reeuq) [reply src/transformers/trainer.py:5107]: Why do we no longer need to shift labels `["labels"][...,1:]`  when getting num_items_in_batch?

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -3643,15 +3650,11 @@ def training_step(
             with amp.scale_loss(loss, self.optimizer) as scaled_loss:
                 scaled_loss.backward()
         else:
-            if num_items_in_batch is not None:
-                if self.compute_loss_func or self.model_accepts_loss_kwargs:
-                    loss *= self.args.gradient_accumulation_steps
```
Comment (reeuq) [inline src/transformers/trainer.py:3648]: I wonder why we don't need `loss *= self.args.gradient_accumulation_steps`

Comment (hiyouga) [inline src/transformers/trainer.py:3648]: I'm confused that the loss is no longer multiplied by the gradient accumulation steps here, because the loss has been multiplied by the data parallel size in https://github.com/huggingface/transformers/pull/34511/files#diff-ed55888e6665791fe92cc8fc0c499da54f4ace6738551cd9a2591881cda076deR3702-R3703

Comment (hiyouga) [reply src/transformers/trainer.py:3648]: I see, it should be solved in https://github.com/huggingface/transformers/pull/35207
